[Chen Lyuting] iP

lyuting47 opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~fast~~ *SUPER FAST* to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

lyuting47 committed:
Added level 1, added *.class to .gitignore

lyuting47 committed:
Added level 2

lyuting47 committed:
Added level 3, implemented Task Class

lyuting47 committed:
Added level 4, extended Task to implement Todo, Event and Deadline

lyuting47 committed:
Added automatic testing

lyuting47 committed:
Added Level 5, implemented several DukeExceptions

lyuting47 committed:
Added Level 6, used ArrayList to store tasks

lyuting47 committed:
Add Level 7, fixed missing stuff from Level 6

lyuting47 committed:
Add Level 8

lyuting47 committed:
Fix bugs for level 7

lyuting47 committed:
Merge branch 'branch-Level-7'

lyuting47 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

lyuting47 committed:
Add MoreOOP and Packages

lyuting47 committed:
Add Gradle

lyuting47 committed:
Add JUnit Tests

lyuting47 committed:
Add JavaDoc

lyuting47 committed:
Improve coding standard

lyuting47 committed:
Add Level 9

lyuting47 committed:
Merge branch 'branch-Level-9'

lyuting47 committed:
Merge branch 'branch-A-CodingStandard'

lyuting47 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

lyuting47 committed:
Add CheckStyle

JonathanWiguna commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,56 @@
+        this.isDone = false;
+    }
+
+    public Task(String description, boolean done) {
```
Shouldn't the boolean here be called isDone instead?

JonathanWiguna commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,191 @@
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
+    private int tasksLength;
```
Since you initialize the ArrayList in both constructor, do we still need tasksLength?

JonathanWiguna commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,191 @@
+        }
+        return result;
+    }
+}
```
Great that you have specific exceptions for different problems!

JonathanWiguna commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    public void printSpacer() {
+        System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
```
I think you should break this line into two or make it shorter

JonathanWiguna commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,117 @@
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
```
Good import statements arrangements

JonathanWiguna commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,117 @@
+
+public class DukeTest {
+
+    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+    private final ByteArrayOutputStream errContent = new ByteArrayOutputStream();
+    private final PrintStream originalOut = System.out;
+    private final PrintStream originalErr = System.err;
```
Since these are final, i.e. constants, shouldn't they be in all uppercase with underscore to separate the words?

JonathanWiguna reviewed (commented):
Overall looks good! Just some minor changes requested.

lyuting47 committed:
Add GUI for Duke

Francis-Tan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,117 @@
+ */
+public class Parser {
+
+    private static boolean startsWith(String command, String prefix) {
+        String[] stuff = command.split(" ");
+        return stuff[0].equals(prefix);
+    }
```
Perhaps use the string startsWith method instead?

Francis-Tan commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,103 @@
+ */
+public class Ui {
+
+    private boolean cont = true;
```
Perhaps rename this to isActive

Francis-Tan commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Test
+    public void ui_printTaskList_oneTaskPrinted() {
```
I like how test cases have helpful names, maybe it'd be better to split them into different classes?

Francis-Tan replied:
Perhaps split the classes in duke into different packages too

Francis-Tan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,191 @@
+            throw new EmptyTodoException();
+        }
+
+        String result = line.substring(5);
+        Todo t = new Todo(result);
```
Perhaps some variable names could be made more explicit for clarity, such as renaming result to desc

Francis-Tan reviewed (commented):
Good job! I found your code clean, readable and easy to follow

lyuting47 committed:
Add Level 10

lyuting47 committed:
Add GUI to Duke

A GUI implemented using JavaFX is now used
instead of CLI.

The GUI gives users a better experience as there
are more visuals and makes the app look more
professional.

Let's
* present the app as a chatbot using JavaFX
* use a picture of king to represent the user,
and a bishop to represent Duke
* displays command inputs and Duke's outputs
as speech bubbles by the king and bishop

lyuting47 committed:
Test

lyuting47 committed:
Merge pull request #2 from lyuting47/branch-test

Test

lyuting47 committed:
Improve code quality

lyuting47 committed:
Merge pull request #3 from lyuting47/branch-A-CodeQuality

Improve code quality

lyuting47 committed:
Add assertions

lyuting47 committed:
Merge pull request #4 from lyuting47/branch-A-Assertions

Add assertions

lyuting47 committed:
Add update function

lyuting47 committed:
Merge branch 'branch-BCD-Extension'

lyuting47 committed:
Add Ui.png

lyuting47 committed:
Rename Ui.png to correct format

lyuting47 committed:
Add personality

lyuting47 committed:
Change build.gradle config to build JAR with Launcher

lyuting47 committed:
Rename Ui.png

lyuting47 committed:
Add README

lyuting47 committed:
Update README

lyuting47 committed:
Reformat README more

lyuting47 committed:
Update README.md to include datetime formatting

lyuting47 committed:
Add exit function to README.md

lyuting47 committed:
Fix JUnit tests

lyuting47 committed:
Merge remote-tracking branch 'origin/master'
